[Feat] #41 - 로그인 구현

joho2022 opened the pull request:
## 🌴 작업한 브랜치
- #41 


## ✅ 작업한 내용
<!-- 작업한 내용 적어주세요! -->
- [x] 어댑터 패턴(Adapter)으로 소셜 로그인 구현
- [x] 온보딩 완료 여부에 따라 TabBarView ↔ OnboardingView 분기 처리용 Auth 전용 NavigationPath 관리
- [x] 24시간 기준으로 Onboarding 체크포인트 영속 저장 구현
- [x] 디버깅 모드에서 Keychain·UserDefaults 내 체크포인트 초기화 기능 도입
- [x] 401 응답 시 Keychain의 Refresh Token으로 1회 재시도하는 토큰 갱신 로직 구현
<!--
```
넣고싶은 코드가 있다면 적어주세요
```
-->


## ❗️PR Point
<!-- 부족했던 점 혹은 개선하고 싶은 방향이 있다면 얘기해주세요 -->
PR분량 조절 실패했습니다 ㅎㅎ.. ㅜㅠ
## 핵심 로직인 만큼, 아래 객체들이 저의 의도와 수행하는 역할을 이해하기 쉽도록 설명을 함께 첨부했습니다.

## AuthManager
앱 전역에서 소셜 로그인 흐름을 관리하는 싱글톤 객체
현재 카카오, 애플이지만, 소셜로그인이 추가되는 확장성을 고려하여
어댑터 패턴을 적용하였습니다.

1. 로그인이 시작되면 각 어댑터를 생성하고 로그인 실행하여 인가코드 또는 토큰을 획득
2. 해당 받은 코드 또는 토큰을 가지고 AuthService에 요청합니다. 성공적으로 응답이 오면 온보딩 단계를 결정
3. 토큰 키체인에 저장하고 온보딩 단계는 UserDefaults에 체크포인트 기록
4. DEBUG환경에서 키체인과 체크포인트를 초기화해서 원활한 온보딩 플로우까지 테스트할 수 있도록 의도

## KeychainManager
Keychain을 이용해 액세스 토큰과 리프레시 토큰을 안전하게 CRUD하는 객체
do-catch를 각 메서드에 넣으면 try/catch를 반복해야 해서 코드가 불필요하게 장황해집니다. 

반면 Result 타입을 반환하면 오류 매핑을 한곳에서 처리하고, 
호출부는 switch나 if문으로 성공·실패만 간단히 분기할 수 있으므로, 호출 결과를 Result 타입으로 반환하도록 구현했습니다.

## TokenRefresher
actor로 구현되어 401 에러 시 중복 없는 토큰 재발급(refreshTask)을 보장하고, 
성공 시 새 액세스 토큰을 Keychain에 저장하는 동시성 안전 로직을 수행하는 객체

최초 호출에만 service.refresh()로 네트워크 요청을 보내고, 
이후 요청은 기존 refreshTask 결과를 재사용해 불필요한 중복 호출을 방지

- 왜 actor? 혹시나 각 API호출이 여러 곳에서 동시에 접근해도 한 번에 하나의 작업만 처리하여 안전하게 관리
- 왜 refreshTask 캐싱하는가? 토큰 재발급 결과를 일관되게 전달하기 위해서
- 왜 BaseService에서 Task.detached ? 콜백 클로저는 async가 아니기 때문에, await을 사용하기 위해 내부에서 아에 독립으로 비동기가 실행하도록 해서, 실행흐름을 안전하게 보장하기 위해서입니다.

## OnboardingManager
기능 명세서에서 24시간 기준으로 체크포인트 관리하기 때문에,
24시간(86,400초)이 경과하면 체크포인트를 자동 만료 처리하기 위해
Date().timeIntervalSince(lastActive) > 24 * 60 * 60 조건으로 분기 잡음.

## RootView
인증관련 경로와 탭바 내 화면 경로를 독립적으로 관리

많은 의견 주시면 감사하겠습니다.
<!--
```
넣고싶은 코드가 있다면 적어주세요
```
-->


## 📸 스크린샷
<!-- 스크린 샷을 첨부해주세요 -->
|기능|카카오 로그인|애플 로그인|
|:--:|:--:|:--:|
|GIF|<img src = "https://github.com/user-attachments/assets/dcbca1c0-64b4-4435-8b68-6c1148bc0415" width ="250">|<img src = "https://github.com/user-attachments/assets/d9a83956-0eb9-4f03-9196-7bdd7ef08a60" width ="250">|



## 📟 관련 이슈
- Resolved: #41

joho2022 committed:
[Add] #41 - Login Button Assets 추가

joho2022 committed:
[Del] #41 - .gitkeep 삭제

joho2022 committed:
[Del] #41 - .gitkeep 삭제

joho2022 committed:
[Feat] #41- 인증 응답 및 로그인 요청 DTO 구현

joho2022 committed:
[Feat] #41 - Kakao Login 설정 추가

joho2022 committed:
[Feat] #41 - Kakao SDK 초기화 및 딥링크 처리 구현

joho2022 committed:
[Feat] #41 - User 모델 및 UserRole enum 구현

joho2022 committed:
[Feat] #41 - AuthManager 및 AuthError 구현

joho2022 committed:
[Feat] #41 - KeychainManager 구현

joho2022 committed:
[Feat] #41 - KakaoLoginAdapter 및 SocialLoginService 구현

joho2022 committed:
[Feat] #41 - AuthAPI 및 AuthService 구현

joho2022 committed:
[Feat] #41 - TokenRefresher 구현

joho2022 committed:
[Feat] #41 - 재발급 토큰 재시도 1회 구현 및 BaseTargetType header 처리

joho2022 committed:
[Merge] #41 - Conflicts 해결

# Conflicts:
#	Napzak-iOS/Napzak-iOS.xcodeproj/project.pbxproj

joho2022 committed:
[Feat] #41 - 로그인 인증 로직 및 OnboardingStep 분기 처리 추가

joho2022 committed:
[Feat] #41 - Apple Sign In Entitlement 활성화

joho2022 committed:
[Feat] #41 - Apple 어댑터 구현

joho2022 committed:
[Feat] #41 - Kakao 어댑터 구현

joho2022 committed:
[Feat] #41 - MockAuthService 구현

joho2022 committed:
[Refactor] #41 - BaseService request 메서드 명시화

joho2022 committed:
[Del] #41 - Auth User 모델 삭제

joho2022 committed:
[Feat] #41 - 유효기간 기반 Onboarding 체크포인트 저장 로직 추가

joho2022 committed:
[Fix] #41 - AuthData.nickname을 옵셔널로 변경

joho2022 committed:
[Feat] #41 - 소셜 로그인 타입별 경로·플랫폼·쿼리키 정의 기능 추가

joho2022 committed:
[Feat] #41 - 플랫폼 파라미터와 소셜 로그인 경로·쿼리 동적 로직 추가

joho2022 committed:
[Feat] #41 - RootView 및 AuthNavigationRouter 구현

joho2022 committed:
[Setting] #41 - Xcode 프로젝트 빌드 설정 갱신

joho2022 committed:
[Chore] #41 - 누락된 detached 추가

joho2022 committed:
[Chore] #41 - 인증관련 경로 초기화로 매끄러운 화면 전환 개선

OneTen19 commented on Napzak-iOS/Napzak-iOS/Application/Napzak_iOSApp.swift:
```diff
@@ -6,13 +6,25 @@
+    
+    init() {
+        let kakaoAppKey = Bundle.main.infoDictionary?["KAKAO_APP_KEY"] as? String ?? ""
+        KakaoSDK.initSDK(appKey: kakaoAppKey)
+    }
+    
     var body: some Scene {
         WindowGroup {
-            NZTabBarView()
-                .environmentObject(NavigationRouter())
+            RootView()
+                .onOpenURL { url in
+                    if AuthApi.isKakaoTalkLoginUrl(url) {
+                        _ = AuthController.handleOpenUrl(url: url)
+                    }
+                }
```
앱 시작할 때 KEY있는지 확인하고 있으면 RootView로 호출할 때 로그인 시켜주고 없으면 로그인 뷰로 넘어가는 방식인감...

joho2022 replied:
Info.plist에서 읽어온 앱 키를 SDK에 등록합니다
앱 키는 카카오 디벨로퍼에서 확인 가능하구,

onOpenURL은 카카오 로그인 성공이 되면, 다시 납작마켓 앱으로 돌아가기 위한 주소를 설정하는 로직이라고 
이해하시면 됩니다. 

초기화 로직은 카카오 SDK 문서 예시 그대로 사용했습니다.

로그인 정보를 바탕으로 
즉, 우리 앱에서는 키체인 내에서 토큰 유무에 따라 분기잡는건
RootView 내부에 AuthManager에서 이뤄져서  결정하도록 구현했어요

OneTen19 commented on Napzak-iOS/Napzak-iOS/Global/Core/Auth/SocialLogin/Adapters/KakaoLoginAdapter.swift:
```diff
@@ -0,0 +1,87 @@
+    
+    @MainActor
+    private func loginWithKakaoTalk() async -> Result<String, AuthError> {
+        return await withCheckedContinuation { continuation in
+            UserApi.shared.loginWithKakaoTalk { oauthToken, error in
+                if let error = error {
+                    Self.logger.error("KakaoTalk login failed: \(error.localizedDescription)")
+                    continuation.resume(returning: .failure(.loginFailed(service: .kakao)))
+                    return
+                }
+                
+                guard let authCode = oauthToken?.accessToken else {
+                    Self.logger.error("KakaoTalk login failed: missing accessToken")
+                    continuation.resume(returning: .failure(.loginFailed(service: .kakao)))
+                    return
+                }
+                
+                Self.logger.info("KakaoTalk login succeeded")
+                continuation.resume(returning: .success(authCode))
+            }
+        }
```
카카오 로그인도 withCheckedContinuation 로 wrapping해줘야하나보네유

joho2022 replied:
카카오 SDK 에서 콜백클로저로 구현되어 있어서 
async / await 으로 통일해서 사용할려고 변환하였습니다.

OneTen19 reviewed (approved):
고생하셨습니다ㅏ
코드 매우매우 어렵군
진짜 레전드 천재만재

lalaurrel reviewed (approved):
LGTM
고생하셨습니다 호근핑..
코드에서 눈물이 보이는 건 기분탓이겠죠

Johyerin commented on Napzak-iOS/Napzak-iOS/Global/Core/Auth/AuthManager.swift:
```diff
@@ -0,0 +1,120 @@
+    private let keychain = KeychainManager.shared
+    private let onboardingManager = OnboardingManager.shared
+    private let authService: AuthServiceProtocol
+    
```
P2

NetworkService 파일 내부에서 선언하여 사용하는 게 좋을 것 같습니당

Johyerin commented on Napzak-iOS/Napzak-iOS/Global/Core/Auth/KeychainManager.swift:
```diff
@@ -0,0 +1,114 @@
+        }
+    }
+    
+    @discardableResult
+    func updateAccessToken(_ token: String) -> Result<Void, AuthError> {
```
옹 이런 속성도 있군요?

Johyerin commented on Napzak-iOS/Napzak-iOS/Global/Core/Auth/SocialLogin/Adapters/KakaoLoginAdapter.swift:
```diff
@@ -0,0 +1,87 @@
+    
+    @MainActor
+    func login() async -> Result<String, AuthError> {
+        if UserApi.isKakaoTalkLoginAvailable() {
+            return await loginWithKakaoTalk()
+        } else {
+            return await loginWithKakaoAccount()
+        }
```
웹로그인도 고려한 고능호사마

Johyerin commented on Napzak-iOS/Napzak-iOS/Global/Core/Auth/TokenRefresher.swift:
```diff
@@ -0,0 +1,60 @@
+import Foundation
+import os
+
+actor TokenRefresher {
+    static let shared = TokenRefresher()
```
Intercepter를 활용한 방식만 알고있었는데 배워갑니다ㅏ 🥹🥹🥹

joho2022 replied:
Intercepter 몰라서 해당 방법으로 한거였는데, Intercepter가 더 보기 편한거 같네요 나중에 적용해야할 것 같네요

Johyerin commented on Napzak-iOS/Napzak-iOS/Network/Auth/AuthAPI.swift:
```diff
@@ -0,0 +1,61 @@
+//
+
+import Moya
+import Foundation
+
```
P2

Foundation은 지워줘도 될듯용

Johyerin commented on Napzak-iOS/Napzak-iOS/Network/Auth/AuthService.swift:
```diff
@@ -0,0 +1,22 @@
+//
+//  Created by 조호근 on 5/1/25.
+//
+
+import Foundation
```
P2

여기두

Johyerin commented on Napzak-iOS/Napzak-iOS/Network/Auth/MockAuthService.swift:
```diff
@@ -0,0 +1,27 @@
+//  Created by 조호근 on 5/3/25.
+//
+
+import Foundation
+
+final class MockAuthService: AuthServiceProtocol {
+    func login(type: SocialLoginType, authorizationCode: String) async -> Result<AuthResponseDTO, NetworkError> {
```
P2

추후에 사용하지 않을 파일이라면 지워주시조

Johyerin commented on Napzak-iOS/Napzak-iOS/Global/Core/Auth/TokenRefresher.swift:
```diff
@@ -0,0 +1,60 @@
+    static let shared = TokenRefresher()
+
+    private let logger = Logger(subsystem: Bundle.main.bundleIdentifier ?? "Napzak", category: "Auth.Refresh")
+    
+    private let service: TokenRefreshServiceProtocol = TokenRefreshService()
```
P2

요것도 `NetworkService`에서 한번에 관리하는 것이 좋을까나...

Johyerin reviewed (approved):
### LGTM
소셜로그인에 키체인에 리프레셔까지 레전드 고생핑입니다

joho2022 committed:
[Chore] #41 - 피드백 개선

joho2022 committed:
[Merge] #41 - Conflicts 해결

# Conflicts:
#	Napzak-iOS/Napzak-iOS.xcodeproj/project.pbxproj
#	Napzak-iOS/Napzak-iOS/Presentation/TabBar/View/NZTabBarView.swift
